### ISSUE TITLE
Confusing debug message when decoding associated values.

### ISSUE THREAD

**Issue by notcome** [groue/GRDB.swift#836]
### What did you do?

I am trying to decode the following types:

```swift
struct DocumentContent: Equatable, Codable, Identifiable, FetchableRecord, PersistableRecord {
    // Irrelevant code...
    static let blocks = hasMany(StoredBlock.self)
    // Irrelevant code...

struct StoredBlock: Equatable, Codable, Identifiable, FetchableRecord, PersistableRecord {
    // Irrelevant code...

struct AggregatedDocument: Decodable, FetchableRecord {
    var index: DocumentContent
    var blocks: [StoredBlock]

    static let request = DocumentContent.including(all: DocumentContent.blocks)
}
```

This code does not work, because by default GRDB will name the associated values as `storedBlocks`. The correct fix is to have

```swift
static let blocks = hasMany(StoredBlock.self, key: "blocks")
```

As the relation specification.

### What happened instead?

However, the error message is very misleading, in my opinion. I am getting:

```
No such key: index
```

I tracked it down to `FetchableRecord+Decodable.swift`:

```swift
            if let decodedRootKey = decodedRootKey {
                throw DecodingError.keyNotFound(decodedRootKey, DecodingError.Context(
                    codingPath: codingPath,
                    debugDescription: "No such key: \(decodedRootKey.stringValue)"))
            } else {
                decodedRootKey = key
                return try decode(type, fromRow: row, codingPath: codingPath + [key])
            }
```

You can search the key word "no such key" to find it.

The `decodedRootKey` was set to `index` in my case. It is correctly decoded. In other words, this key exists. The reason causing my error was because GRDB decoded my blocks as "storedBlocks" and yet `Decodable` expects a "blocks".

So, shouldn't the debug description be "No such key: \(key.stringValue)"?

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** Revision ededd86.
**Installation method:** SPM
**Xcode version:** Xcode 12.
**Swift version:** 5.3
**Platform(s) running GRDB:** iOS simulator.
**macOS version running Xcode:** Catalina.

**Comment by groue** [comment]
Hello @notcome, thanks for reporting this! I think the error message contains the wrong key indeed.

**Comment by groue** [comment]
@notcome, I'm sorry but it's more complicated than that. I need a reproducible case in order to fully understand what is happening.

Would you please provide a sample code like this one below?

```swift
import GRDB

// Record definitions
struct DocumentContent { ... }

// Create an in-memory database
let dbQueue = DatabaseQueue()

try dbQueue.write { db in
    // Setup schema
    try db.create(table...)

    // Insert rows
    try DocumentContent(...).insert(db)

    // Reproduce the issue
    let request = ...
    _ = try request.fetchAll(db)
}
```

Thank you very much!

**Comment by notcome** [comment]
I do not have a good Internet connection so creating a completely fresh Swift packages loading GRDB (> 100MB) is not possible. So forgive me for writing it as a test case:

```swift
import XCTest
import GRDB

struct Author: Codable, FetchableRecord, PersistableRecord {
    var name: String

    static let name: Column = Column(CodingKeys.name)
    static let books = hasMany(PublishedBooks.self)

    static func createTable(in db: Database) throws {
        try db.create(table: databaseTableName) { td in
            td.column(name.name, .text).primaryKey()
        }
    }
}

struct PublishedBooks: Codable, FetchableRecord, PersistableRecord {
    var bookName: String
    var author: String

    static let bookName: Column = Column(CodingKeys.bookName)
    static let author: Column = Column(CodingKeys.author)

    static func createTable(in db: Database) throws {
        try db.create(table: databaseTableName) { td in
            td.column(bookName.name, .text).primaryKey()
            td.column(author.name, .text)
                .notNull()
                .references(Author.databaseTableName)
        }
    }
}

struct JoinedFetch: Codable, FetchableRecord {
    var author: Author
    // Use "books" instead of "publishedBooks".
    var books: [PublishedBooks]

    static let request = Author.including(all: Author.books)
}

class ReproducingErrorTests: XCTestCase {
    func testKeyNotFound() throws {
        let dbQueue = DatabaseQueue()
        try dbQueue.write { db in
            try Author.createTable(in: db)
            try PublishedBooks.createTable(in: db)
            try Author(name: "Alice").save(db)
            try PublishedBooks(bookName: "X", author: "Alice").save(db)
            try PublishedBooks(bookName: "Y", author: "Alice").save(db)
            try PublishedBooks(bookName: "Z", author: "Alice").save(db)
        }
        try dbQueue.read { db in
            let rows = try JoinedFetch.fetchAll(db, JoinedFetch.request)
            print(rows)
        }
    }
}
```

Here comes the error message:

```
2020-09-18 15:09:56.225432+0800 xctest[19148:356792] Fatal error: 'try!' expression unexpectedly raised an error: Swift.DecodingError.keyNotFound(CodingKeys(stringValue: "author", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No such key: author", underlyingError: nil)): file <some-path>/GRDB.swift/GRDB/Record/FetchableRecord+Decodable.swift, line 7
```

**Comment by notcome** [comment]
Change `books` to `publishedBooks` in `JoinedFetch` solves this issue.

**Comment by groue** [comment]
👍 This is great, @notcome, thank you very much. I'll have a look at it.

**Comment by groue** [comment]
The error message for your reproducible case will now be `No such key: author or books`. It is not a great improvement, I agree, but it's quite complicated to do better.

**Comment by groue** [comment]
The improved diagnostic has shipped in GRDB 5.0.0
